Added step-back strategy and match checking

EToreo opened the pull request:
Retries are now rate-limited with a step-back strategy to avoid hammering.

Match option now allows client to only retry if rejection matches one of list of matches (either string compare or instanceof).

mickhansen commented on index.js:
```diff
@@ -13,30 +13,69 @@ module.exports = function(callback, options) {
+    timeout: options.timeout || undefined,
+    match: options.match || [],
+    stepbackBase: options.stepbackBase === undefined ? 1.2 : options.stepbackBase,
+    stepbackExponent: options.stepbackExponent || 1.5
```
`backoff` might be more fitting.

mickhansen replied:
If the default for exponential backoff is not 0 this will likely have to be a major bump (but that's probably ok)

mickhansen commented on index.js:
```diff
@@ -13,30 +13,69 @@ module.exports = function(callback, options) {
+  
+  // Massage match option into array so we can blindly treat it as such later
+  if(!Array.isArray(options.match))
+    options.match = [options.match];
```
Use a single line or wrap in brackets, please also add a space after if. (goes for the rest of the changes too please :))

mickhansen commented on index.js:
```diff
@@ -13,30 +13,69 @@ module.exports = function(callback, options) {
+        debug('Delay not in use');
+        module.exports(callback, options).then(resolve).catch(reject);
+      }
+
```
```
if {

} else {

}
```

Please :)

mickhansen commented on index.js:
```diff
@@ -13,30 +13,69 @@ module.exports = function(callback, options) {
+        debug('Delay set to %s', options.stepbackBase);
+        stepbackTimeout = setTimeout(function() {
+          module.exports(callback, options).then(resolve).catch(reject);
+        }, options.stepbackBase);
```
Since bluebird is currently a dependency it might be better to just use `Promise.delay`

mickhansen commented on test/bluebird.test.js:
```diff
@@ -94,10 +100,189 @@ describe('bluebird', function () {
+        stepbackBase: 1,
+        match: soRejected
+      })).to.eventually.be.equal(soResolved).then(function () {
+        expect(count).to.equal(4);
```
Since you're testing a call count it might be better to just introduce a sinon stub here. (Goes for all tests that test call counts)

mickhansen commented on test/bluebird.test.js:
```diff
@@ -94,10 +100,189 @@ describe('bluebird', function () {
+    });
+
+    it('should continue retry while error is instanceof match in array', function () {
+        return expect(retry(function () {
```
same indentation issue on all tests :)

mickhansen commented on test/bluebird.test.js:
```diff
@@ -94,10 +100,189 @@ describe('bluebird', function () {
+        expect(count).to.equal(10);
+        expect(moment().diff(startTime)).to.be.above(1200);
+      });
+    });
```
Would also be easier to test with a stub on Promise.delay and just verifying that the call values to it are as expected

EToreo commented:
@mickhansen Thanks for all the feedback.  Most of has been absorbed and updated.  I have a few questions about some of your notes, though.

I changed "stepback" to "backoff" (no idea why I didn't call it "backoff" to start with - going to blame it on my cold).

I think its important that the default behavior have "backoff" enabled.  Mindlessly hammering anything is probably a bad idea.  If the future we could add different back-off strategies (none, linear, double, etc. - in fact double might be a better default behavior than exponential).

I used setTimeout for the delay so it can be canceled if `timeout` is activated.  I didn't see a way to do that with bluebird's delay.

I'm not sure how a sinon stub would work here.  I would need to see an example in a similar test.

I'm not sure what the indentation issues are on the tests... I think the tests I added follow the syntax of the other tests.

mickhansen commented:
@EToreo i agree that backoff should be default, was just noting to myself to bump major :)
Hmm, just looks like the first line of each test is indented 2 spaces too far.

I'll comment with an example of a stub.

mickhansen commented on test/bluebird.test.js:
```diff
@@ -94,10 +100,189 @@ describe('bluebird', function () {
+      })).to.eventually.be.equal(soResolved).then(function () {
+        expect(count).to.equal(4);
+      });
+    });
```
I realize my original tests use count aswell, but i've gotten wiser since then.
We should add the packages `sinon`, `sinon-chai` and `sinon-as-promised`.

This test could then be refactored to:

``` js
var callback = this.sinon.stub();
callback.resolves(soResolved);
callback.onCall(4).rejects(soRejected);
return expect(retry(callback)).to.eventually.equal(soResolved).then(function () {
  expect(callback.callCount).to.equal(4);
});
```

EToreo commented:
@mickhansen I think the indenting is correct.  The tests I added are wrapped in `describe` blocks - one for `match` and the other for `backoff`.  This could explain the extra two spaces.

I will try to update all the tests with the stub.

mickhansen commented on test/bluebird.test.js:
```diff
@@ -94,10 +100,189 @@ describe('bluebird', function () {
+    });
+    
+    it('should continue retry while error is instanceof match', function () {
+        return expect(retry(function () {
```
Do you not see the extra indentation here in the diff? :)

EToreo replied:
Now I do :)

EToreo commented:
@mickhansen  Trying to convert one of the simpler tests to use sinon gives some unexpected (to me) results:

  1) bluebird should resolve if resolved before hitting max:
      AssertionError: expected 5 to equal 4

``` javascript
  it('should resolve if resolved before hitting max', function () {
    var callback = sinon.stub();
    callback.rejects(soRejected);
    callback.onCall(4).resolves(soResolved);
    return expect(retry(callback, {max: 10, backoffBase: 0})).to.eventually.equal(soResolved).then(function () {
      expect(callback.callCount).to.equal(4);
    });
  });
```

It seems callback was called 5 times (confirmed by the debug output), even though `onCall(4)` was used.  Is this written incorrectly?  Do I have some fundamental misunderstanding of how `stub` or `onCall` works?

mickhansen commented:
@EToreo Oh, onCall is zero indexed, so i guess that should probably be `onCall(3)` (http://sinonjs.org/docs/#stub-onCall). My mistake.

EToreo commented:
@mickhansen Bummer.  That makes the test read poorly because of the number miss-match.

mickhansen commented:
@EToreo agreed, i usually use `onFirstCall()`/`onThirdCall` but that only goes up to three calls.

EToreo commented:
@mickhansen OK, I think that's everything we talked about.  Let me know.

mickhansen commented on index.js:
```diff
@@ -13,30 +13,66 @@ module.exports = function(callback, options) {
+      // Only continue retrying if match is zero length, or we find a match
+      var continueTrying = options.match.length === 0;
+      options.match.forEach(function(match) {
+        if (typeof match === "string" && match === err.toString())
```
Might want to check against `err.message` instead, i believe that's how bluebird does it for catch.

mickhansen commented on index.js:
```diff
@@ -13,30 +13,66 @@ module.exports = function(callback, options) {
-    timeout: options.timeout || undefined
+    timeout: options.timeout || undefined,
+    match: options.match || [],
+    backoffBase: options.backoffBase === undefined ? 1.2 : options.backoffBase,
```
The default backup is 1.2ms? Should perhaps be 120 or just 100?

EToreo commented:
@mickhansen Sorry for the delay.  I was sick and this weekend was a holiday.

EToreo commented:
@mickhansen Can we get this pulled?  I need it for a pull request I'm working on for Sequelize.

mickhansen commented:
`v2.0.0`

EToreo commented:
@mickhansen Thanks!
